File: backend/app/schemas/director_pack.py
"""
Director Pack Schema Definitions

DirectorPack: Compressed coaching rules for real-time Gemini Live sessions.

Philosophy:
- VDG = Brain (SoR), Pack = Script (compressed rules)
- DNA Invariants = What to KEEP (불변 규칙)
- Mutation Slots = What CAN change (가변 요소)
- Forbidden Mutations = What to NEVER do (금지 규칙)

License: arkain.info@gmail.com (Gemini Enterprise)
"""
from typing import Dict, List, Optional, Any, Tuple
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# =============================================================================
# Enums
# =============================================================================

class RulePriority(str, Enum):
    """Priority levels for rules."""
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class InvariantType(str, Enum):
    """Types of DNA invariants."""
    TIMING = "timing"
    COMPOSITION = "composition"
    ENGAGEMENT = "engagement"
    AUDIO = "audio"
    NARRATIVE = "narrative"
    TECHNICAL = "technical"


class SlotType(str, Enum):
    """Types of mutation slots."""
    STYLE = "style"
    TONE = "tone"
    PACING = "pacing"
    COLOR = "color"
    MUSIC = "music"
    TEXT = "text"


class Severity(str, Enum):
    """Severity levels for forbidden mutations."""
    CRITICAL = "critical"
    MAJOR = "major"
    MINOR = "minor"


# =============================================================================
# Source Reference & Evidence
# =============================================================================

class SourceRef(BaseModel):
    """Reference to source evidence."""
    source_type: str = Field(description="'vdg', 'metric', 'entity', 'frame'")
    source_id: str = Field(description="ID of the source")
    timestamp: Optional[float] = Field(default=None, description="Relevant timestamp")
    confidence: float = Field(default=1.0, ge=0.0, le=1.0)


# =============================================================================
# DNA Invariants (What to KEEP)
# =============================================================================

class RuleSpec(BaseModel):
    """Specification for a coaching rule."""
    operator: str = Field(description="'eq', 'gt', 'lt', 'gte', 'lte', 'between', 'in', 'pattern', 'exists', '~='")
    value: Any = Field(description="Target value or range")
    tolerance: Optional[float] = Field(default=None, description="Acceptable deviation")
    unit: Optional[str] = Field(default=None, description="Unit of measurement, e.g. 'sec', 'percent'")
    context_filter: Optional[List[str]] = Field(
        default=None,
        description="Contexts where this rule applies, e.g. ['sequence_start', 'shortform_start']"
    )



class TimeScope(BaseModel):
    """Time scope for rule application."""
    t_start: float = Field(description="Start time in seconds")
    t_end: float = Field(description="End time in seconds")
    relative: bool = Field(default=False, description="If true, relative to video start")


class DNAInvariant(BaseModel):
    """
    단일 불변 규칙 (DNA Invariant)
    
    Example: "Hook must appear within first 2 seconds"
    """
    rule_id: str = Field(description="Unique rule ID, e.g. 'inv_hook_timing_001'")
    rule_type: InvariantType = Field(description="Type of invariant")
    name: str = Field(description="Human-readable name")
    description: str = Field(default="", description="Detailed description")
    
    # Rule specification
    condition: str = Field(description="What to check, e.g. 'hook_start_time'")
    spec: RuleSpec = Field(description="Rule specification")
    
    # Scope & Priority
    time_scope: Optional[TimeScope] = Field(default=None, description="When this applies")
    priority: RulePriority = Field(default=RulePriority.MEDIUM)
    
    # Evidence
    source_refs: List[SourceRef] = Field(default_factory=list)
    confidence: float = Field(default=0.5, ge=0.0, le=1.0)
    
    # Coaching
    coach_line: Optional[str] = Field(default=None, description="What to say when violated")
    coach_line_ko: Optional[str] = Field(default=None, description="Korean version")


# =============================================================================
# Mutation Slots (What CAN change)
# =============================================================================

class MutationSlot(BaseModel):
    """
    가변 요소 슬롯 (Mutation Slot)
    
    Example: "Background music style can be energetic, calm, or dramatic"
    """
    slot_id: str = Field(description="Unique slot ID, e.g. 'slot_music_style'")
    slot_type: SlotType = Field(description="Type of slot")
    name: str = Field(description="Human-readable name")
    description: str = Field(default="")
    
    # Allowed values
    allowed_values: Optional[List[Any]] = Field(default=None, description="Discrete options")
    allowed_range: Optional[Tuple[float, float]] = Field(default=None, description="Numeric range")
    default_value: Optional[Any] = Field(default=None)
    
    # Persona binding
    persona_presets: Dict[str, Any] = Field(default_factory=dict, description="{persona: recommended_value}")
    
    # Evidence
    source_refs: List[SourceRef] = Field(default_factory=list)


# =============================================================================
# Forbidden Mutations (What to NEVER do)
# =============================================================================

class ForbiddenMutation(BaseModel):
    """
    금지 규칙 (Forbidden Mutation)
    
    Example: "Never use vertical video for landscape scenes"
    """
    mutation_id: str = Field(description="Unique mutation ID")
    name: str = Field(description="Human-readable name")
    description: str
    
    # What's forbidden
    forbidden_condition: str = Field(description="What to prevent")
    severity: Severity = Field(default=Severity.MAJOR)
    
    # When it applies
    time_scope: Optional[TimeScope] = Field(default=None)
    
    # Coaching
    coach_line: Optional[str] = Field(default=None, description="Warning message")
    coach_line_ko: Optional[str] = Field(default=None)
    
    # Evidence
    source_refs: List[SourceRef] = Field(default_factory=list)


# =============================================================================
# Checkpoints (Time-based rule activation)
# =============================================================================

class Checkpoint(BaseModel):
    """
    Time-based checkpoint for rule monitoring.
    
    Example: "At t=2s, check if hook has fired"
    """
    checkpoint_id: str = Field(description="Unique checkpoint ID")
    t: float = Field(description="Timestamp in seconds")
    
    # What to check
    check_rule_ids: List[str] = Field(default_factory=list, description="Rules to verify")
    
    # Coaching
    coach_prompt: Optional[str] = Field(default=None, description="What to ask/remind")
    coach_prompt_ko: Optional[str] = Field(default=None)


# =============================================================================
# Policy & Runtime Contract
# =============================================================================

class Policy(BaseModel):
    """Coaching policy settings."""
    interrupt_on_violation: bool = Field(default=True, description="Interrupt for critical violations")
    suggest_on_medium: bool = Field(default=True, description="Suggest for medium violations")
    log_all_checks: bool = Field(default=False)
    language: str = Field(default="ko", description="Primary coaching language")


class RuntimeContract(BaseModel):
    """Runtime settings for Gemini Live session."""
    max_session_sec: int = Field(default=300, description="Max session length")
    checkpoint_interval_sec: float = Field(default=5.0, description="Check interval")
    enable_realtime_feedback: bool = Field(default=True)
    enable_audio_coach: bool = Field(default=True)


# =============================================================================
# Coach Line Templates
# =============================================================================

class CoachLineTemplates(BaseModel):
    """Template library for coaching messages."""
    violation_critical: str = Field(default="⚠️ 중요: {rule_name} 위반. {coach_line}")
    violation_major: str = Field(default="💡 개선점: {rule_name}. {coach_line}")
    violation_minor: str = Field(default="참고: {coach_line}")
    encouragement: str = Field(default="✅ 좋아요! {positive_note}")
    checkpoint_reminder: str = Field(default="⏱️ {t}초 체크포인트: {coach_prompt}")


# =============================================================================
# Scoring
# =============================================================================

class Scoring(BaseModel):
    """Scoring weights for coaching evaluation."""
    weights: Dict[str, float] = Field(default_factory=dict, description="{rule_id: weight}")
    total_possible: float = Field(default=100.0)
    pass_threshold: float = Field(default=70.0)


# =============================================================================
# Pack Metadata
# =============================================================================

class PackMeta(BaseModel):
    """Director Pack metadata."""
    pack_id: str = Field(description="Unique pack ID")
    pattern_id: str = Field(description="Source pattern/content ID")
    version: str = Field(default="1.0.0")
    
    # Source
    source_vdg_id: Optional[str] = Field(default=None, description="Source VDG content_id")
    source_quality_tier: Optional[str] = Field(default=None)
    
    # Compilation
    compiled_at: datetime = Field(default_factory=datetime.utcnow)
    compiled_by: str = Field(default="DirectorCompiler")
    
    # Stats
    invariant_count: int = Field(default=0)
    slot_count: int = Field(default=0)
    forbidden_count: int = Field(default=0)
    checkpoint_count: int = Field(default=0)


# =============================================================================
# Director Pack (Main Output)
# =============================================================================

class DirectorPack(BaseModel):
    """
    Director Pack: Compressed rules for real-time AI coaching.
    
    Used by Gemini Live to provide director-style feedback.
    
    Structure:
    - dna_invariants: Rules that MUST be followed
    - mutation_slots: Elements that CAN vary
    - forbidden_mutations: Things to NEVER do
    - checkpoints: Time-based monitoring points
    """
    # Metadata
    meta: PackMeta
    
    # Core Rules
    dna_invariants: List[DNAInvariant] = Field(default_factory=list)
    mutation_slots: List[MutationSlot] = Field(default_factory=list)
    forbidden_mutations: List[ForbiddenMutation] = Field(default_factory=list)
    checkpoints: List[Checkpoint] = Field(default_factory=list)
    
    # Policy & Config
    policy: Policy = Field(default_factory=Policy)
    runtime_contract: RuntimeContract = Field(default_factory=RuntimeContract)
    
    # Templates & Scoring
    coach_templates: CoachLineTemplates = Field(default_factory=CoachLineTemplates)
    scoring: Optional[Scoring] = Field(default=None)
    
    # Quick access
    @property
    def critical_invariants(self) -> List[DNAInvariant]:
        """Get critical priority invariants."""
        return [i for i in self.dna_invariants if i.priority == RulePriority.CRITICAL]
    
    @property
    def rule_count(self) -> int:
        """Total rule count."""
        return len(self.dna_invariants) + len(self.forbidden_mutations)
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }

File: frontend/src/types/director-pack.ts
/**
 * DirectorPack TypeScript Types
 * 
 * Matches backend/app/schemas/director_pack.py
 * Used for type-safe DirectorPack handling in frontend
 */

// =============================================================================
// Enums
// =============================================================================

export type RulePriority = 'critical' | 'high' | 'medium' | 'low';

export type InvariantType =
    | 'timing'
    | 'composition'
    | 'engagement'
    | 'audio'
    | 'narrative'
    | 'technical';

export type SlotType =
    | 'style'
    | 'tone'
    | 'pacing'
    | 'color'
    | 'music'
    | 'text';

export type Severity = 'critical' | 'major' | 'minor';

// =============================================================================
// Source Reference & Evidence
// =============================================================================

export interface SourceRef {
    source_type: 'vdg' | 'metric' | 'entity' | 'frame';
    source_id: string;
    timestamp?: number;
    confidence: number;
}

// =============================================================================
// Rule Components
// =============================================================================

export interface RuleSpec {
    operator: 'eq' | 'gt' | 'lt' | 'gte' | 'lte' | '<=' | '>=' | 'between' | 'in' | 'exists';
    value: unknown;
    tolerance?: number;
    unit?: string;
    aggregation?: 'mean' | 'median' | 'max' | 'min';
    required_inputs?: string[];
}

export interface TimeScope {
    t_start: number;
    t_end: number;
    relative?: boolean;
}

// =============================================================================
// DNA Invariants (What to KEEP)
// =============================================================================

export interface DNAInvariant {
    rule_id: string;
    rule_type: InvariantType;
    name: string;
    description?: string;

    // Rule specification
    condition: string;
    spec: RuleSpec;

    // Scope & Priority
    time_scope?: TimeScope;
    priority: RulePriority;

    // Evidence
    source_refs?: SourceRef[];
    confidence: number;

    // Coaching
    coach_line?: string;
    coach_line_ko?: string;

    // Additional
    weight?: number;
    tolerance?: 'tight' | 'normal' | 'loose';
    fallback?: string;
    evidence_refs?: string[];
}

// =============================================================================
// Mutation Slots (What CAN change)
// =============================================================================

export interface MutationSlot {
    slot_id: string;
    slot_type: SlotType;
    name: string;
    description?: string;

    // Allowed values
    allowed_values?: unknown[];
    allowed_range?: [number, number];
    default_value?: unknown;

    // Persona binding
    persona_presets?: Record<string, unknown>;

    // Evidence
    source_refs?: SourceRef[];

    // Coaching templates per option
    coach_line_templates?: Record<string, string>;
}

// =============================================================================
// Forbidden Mutations (What to NEVER do)
// =============================================================================

export interface ForbiddenMutation {
    mutation_id: string;
    name: string;
    description: string;

    // What's forbidden
    forbidden_condition: string;
    severity: Severity;

    // When it applies
    time_scope?: TimeScope;

    // Coaching
    coach_line?: string;
    coach_line_ko?: string;

    // Evidence
    source_refs?: SourceRef[];
    evidence_refs?: string[];
}

// =============================================================================
// Checkpoints (Time-based rule activation)
// =============================================================================

export interface Checkpoint {
    checkpoint_id: string;
    t: number;

    // What to check - can be either format
    check_rule_ids?: string[];
    active_rules?: string[];
    t_window?: [number, number];

    // Coaching
    coach_prompt?: string;
    coach_prompt_ko?: string;
    note?: string;
}

// =============================================================================
// Policy & Runtime Contract
// =============================================================================

export interface Policy {
    interrupt_on_violation: boolean;
    suggest_on_medium: boolean;
    log_all_checks?: boolean;
    language: 'ko' | 'en';
    one_command_only?: boolean;
    cooldown_sec?: number;
    barge_in_handling?: 'stop_and_ack' | 'ignore' | 'queue';
    uncertainty_policy?: 'ask_user' | 'skip' | 'default';
}

export interface RuntimeContract {
    max_session_sec: number;
    checkpoint_interval_sec: number;
    enable_realtime_feedback: boolean;
    enable_audio_coach: boolean;
    input_modalities_expected?: string[];
    verification_granularity?: 'frame' | 'window' | 'scene';
    max_instruction_words?: number;
    cooldown_sec_default?: number;
}

// =============================================================================
// Coach Line Templates
// =============================================================================

export interface CoachLineTemplates {
    violation_critical: string;
    violation_major: string;
    violation_minor: string;
    encouragement: string;
    checkpoint_reminder: string;
}

// =============================================================================
// Scoring
// =============================================================================

export interface Scoring {
    weights?: Record<string, number>;
    dna_weights?: Record<string, number>;
    total_possible: number;
    pass_threshold: number;
    risk_penalty_rules?: Array<{
        trigger: string;
        penalty: number;
    }>;
}

// =============================================================================
// Pack Metadata
// =============================================================================

export interface PackMeta {
    pack_id: string;
    pattern_id: string;
    version: string;

    // Source
    source_vdg_id?: string;
    source_quality_tier?: 'gold' | 'silver' | 'bronze' | 'reject';

    // Compilation
    compiled_at: string;
    compiled_by?: string;
    generated_at?: string;
    compiler_version?: string;
    source_refs?: Array<{
        vdg_content_id: string;
        vdg_version: string;
    }>;

    // Stats
    invariant_count: number;
    slot_count: number;
    forbidden_count: number;
    checkpoint_count: number;
}

// =============================================================================
// Director Pack (Main Type)
// =============================================================================

export interface DirectorPack {
    // Metadata
    meta: PackMeta;

    // Version info (alternative formats)
    pack_version?: string;
    pattern_id?: string;
    goal?: string;
    pack_meta?: PackMeta;

    // Core Rules
    dna_invariants: DNAInvariant[];
    mutation_slots: MutationSlot[];
    forbidden_mutations: ForbiddenMutation[];
    checkpoints: Checkpoint[];

    // Policy & Config
    policy: Policy;
    runtime_contract: RuntimeContract;

    // Templates & Scoring
    coach_templates?: CoachLineTemplates;
    scoring?: Scoring;
}

// =============================================================================
// API Response Types
// =============================================================================

export interface DirectorPackResponse {
    success: boolean;
    data?: DirectorPack;
    error?: string;
}

export interface DirectorPackListResponse {
    success: boolean;
    data?: DirectorPack[];
    total?: number;
    error?: string;
}

// =============================================================================
// UI State Types
// =============================================================================

export interface DirectorPackViewerState {
    expandedSections: {
        dna: boolean;
        slots: boolean;
        forbidden: boolean;
        timeline: boolean;
    };
    selectedInvariant?: DNAInvariant;
    selectedSlot?: MutationSlot;
    filterPriority?: RulePriority;
}

// =============================================================================
// Utility Types
// =============================================================================

export type DNAInvariantWithStatus = DNAInvariant & {
    status: 'passed' | 'violated' | 'pending' | 'skipped';
    lastCheckedAt?: string;
    violationCount?: number;
};

export type CheckpointWithProgress = Checkpoint & {
    reached: boolean;
    passedRules: string[];
    failedRules: string[];
};
